controllers/complianceeventsapi/types.go

```diff
@@ -0,0 +1,266 @@
+	if err := ce.Policy.Validate(); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+type Cluster struct {
+	KeyID     int    `db:"id" json:"-" goqu:"skipinsert"`
+	Name      string `db:"name" json:"name"`
+	ClusterID string `db:"cluster_id" json:"cluster_id"` //nolint:tagliatelle
+}
+
+func (c Cluster) Validate() error {
+	errs := make([]error, 0)
+
+	if c.Name == "" {
+		errs = append(errs, fmt.Errorf("%w: cluster.name", errRequiredFieldNotProvided))
+	}
+
+	if c.ClusterID == "" {
+		errs = append(errs, fmt.Errorf("%w: cluster.cluster_id", errRequiredFieldNotProvided))
+	}
+
+	return errors.Join(errs...)
+}
+
+type EventDetails struct {
+	KeyID      int       `db:"id" json:"-" goqu:"skipinsert"`
+	ClusterID  int       `db:"cluster_id" json:"-"`
+	PolicyID   int       `db:"policy_id" json:"-"`
+	Compliance string    `db:"compliance" json:"compliance"`
+	Message    string    `db:"message" json:"message"`
+	Timestamp  time.Time `db:"timestamp" json:"timestamp"`
+	Metadata   JSONMap   `db:"metadata" json:"metadata,omitempty"`
+	ReportedBy *string   `db:"reported_by" json:"reported_by,omitempty"` //nolint:tagliatelle
+}
+
+func (e EventDetails) Validate() error {
+	errs := make([]error, 0)
+
+	if e.Compliance == "" {
+		errs = append(errs, fmt.Errorf("%w: event.compliance", errRequiredFieldNotProvided))
+	} else {
+		switch e.Compliance {
+		case "Compliant", "compliant", "NonCompliant", "noncompliant":
+		default:
+			errs = append(errs, fmt.Errorf("%w: event.compliance should be Compliant or NonCompliant, got %v",
+				errInvalidInput, e.Compliance))
+		}
+	}
+
+	if e.Message == "" {
+		errs = append(errs, fmt.Errorf("%w: event.message", errRequiredFieldNotProvided))
+	}
+
+	if e.Timestamp.IsZero() {
+		errs = append(errs, fmt.Errorf("%w: event.timestamp", errRequiredFieldNotProvided))
+	}
+
+	return errors.Join(errs...)
+}
+
+type ParentPolicy struct {
+	KeyID      int            `db:"id" json:"-" goqu:"skipinsert"`
+	Name       string         `db:"name" json:"name"`
+	Categories pq.StringArray `db:"categories" json:"categories,omitempty"`
+	Controls   pq.StringArray `db:"controls" json:"controls,omitempty"`
+	Standards  pq.StringArray `db:"standards" json:"standards,omitempty"`
+}
+
+func (p ParentPolicy) Validate() error {
+	if p.Name == "" {
+		return fmt.Errorf("%w: parent_policy.name", errRequiredFieldNotProvided)
+	}
+
+	return nil
+}
+
+func (p ParentPolicy) key() parentPolicyKey {
+	cats := make([]string, len(p.Categories))
+	for i, c := range p.Categories {
+		cats[i] = fmt.Sprintf("%q", c) // double-quoted and safely escaped with Go syntax
+	}
+
+	ctrls := make([]string, len(p.Controls))
+	for i, c := range p.Controls {
+		ctrls[i] = fmt.Sprintf("%q", c) // double-quoted and safely escaped with Go syntax
+	}
+
+	stands := make([]string, len(p.Standards))
+	for i, c := range p.Standards {
+		stands[i] = fmt.Sprintf("%q", c) // double-quoted and safely escaped with Go syntax
+	}
+
+	return parentPolicyKey{
```

Comment:
I think this or similar would be more efficient as a key.
```go
fmt.Sprintf("%s;%#v;%#v;%#v", p.Name, p.Categories, p.Controls, p.Standards)
```

Reply:
`%#v` was giving a difference between nil and empty, but `%v` worked fine, FYI.